### PR TITLE
AK: Allow variants that may contain Empty to be compared for equality

### PR DESCRIPTION
**Pull request by shannonbooth**

### CODE REVIEW

File: AK/Variant.h [hunk 1]
```diff
@@ -219,6 +219,8 @@ namespace AK {
 struct Empty {
 };
 
+constexpr bool operator==(Empty, Empty) { return true; }
```
Comment (Hendiadyoin1) [inline AK/Variant.h:222]: Does `=default` work as well?

Comment (Hendiadyoin1) [reply AK/Variant.h:222]: Not sure what is preferable here though...

Comment (shannonbooth) [reply AK/Variant.h:222]: Sure - I'll try it

rebuild incoming :(

**Comment by timschumi** [comment]
Is this actually in use anywhere (apart from the newly added tests)? I was going to add it together with what would be the first usage in #18940, but now I'm unsure if I overlooked something.

**Comment by shannonbooth** [comment]
> Is this actually in use anywhere (apart from the newly added tests)? I was going to add it together with what would be the first usage in #18940, but now I'm unsure if I overlooked something.

Not yet - I'm actively working on implementing https://url.spec.whatwg.org/#concept-host-serializer where this was required. Was surprised it didn't work, - and since it was a oneliner patch figured I would just submit it standalone since implementing that was already touching a bit of everywhere. Fair point though, happy to drop this and wait for my/your usecase to get merged in first.

File: AK/Variant.h (hunk 1 above)
Comment (alimpfard) [inline AK/Variant.h:222]: Why is this outside the struct itself?

**Review by alimpfard — COMMENTED** [review]
Commit message seems very misleading to me, instead might I suggest:

```
AK: Allow testing Empty instances for equality

This also makes it possible to compare `Variant<Empty, Ts...>` objects if operator== exists for all Ts.
```

File: Tests/AK/TestVariant.cpp [hunk 2]
```diff
@@ -273,3 +273,26 @@ TEST_CASE(type_list_specialization)
...
+        MyVariant variant1 = 1;
+        MyVariant variant2 = 1.5f;
+        EXPECT_NE(variant1, variant2);
+    }
+
+    {
+        MyVariant variant1 = 1;
+        MyVariant variant2;
+        EXPECT_NE(variant1, variant2);
+    }
```
Comment (gmta) [inline Tests/AK/TestVariant.cpp:297]: Shouldn't we have a test for comparing two empty variants as well?

Comment (shannonbooth) [reply Tests/AK/TestVariant.cpp:297]: Yeah, that makes sense :) Done!

File: AK/Variant.h (hunk 1 above)
Comment (shannonbooth) [reply AK/Variant.h]: No good reason

Changed to Hendiadyoin1's suggestion for this

**Pull request merged**